File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/watch/WatchSync.kt
package com.crispy.tv.domain.watch

enum class WatchSyncConnection {
    DISCONNECTED,
    CONNECTING,
    CONNECTED,
}

data class WatchSyncState(
    val profileId: String,
    val connection: WatchSyncConnection,
    val isSurfaceVisible: Boolean,
)

/**
 * The surface an invalidation targets, sent by the server on the `watch_changed`
 * SSE event so the client can route each invalidation to the right refetch.
 * [Unknown] is a forward-compat sentinel for kinds a new server may emit that
 * this client does not yet understand — the reducer ignores them.
 */
sealed interface WatchSyncKind {
    data object ContinueWatching : WatchSyncKind
    data object History : WatchSyncKind
    data object Watchlist : WatchSyncKind
    data object Ratings : WatchSyncKind
    data object Home : WatchSyncKind
    data object Unknown : WatchSyncKind

    companion object {
        /** A legacy server omits `kind`; treat it as [ContinueWatching]. */
        fun fromRaw(raw: String?): WatchSyncKind =
            when (raw?.trim()?.lowercase()) {
                "continue_watching" -> ContinueWatching
                "history" -> History
                "watchlist" -> Watchlist
                "ratings" -> Ratings
                "home" -> Home
                null, "" -> ContinueWatching
                else -> Unknown
            }
    }
}

sealed interface WatchSyncEvent {
    data object SurfaceBecameVisible : WatchSyncEvent
    data object SurfaceHidden : WatchSyncEvent
    data object ConnectionOpened : WatchSyncEvent
    data object ConnectionClosed : WatchSyncEvent
    data object MaxDurationElapsed : WatchSyncEvent
    data class InvalidationReceived(
        val profileId: String,
        val kind: WatchSyncKind,
        val atMs: Long,
    ) : WatchSyncEvent
}

sealed interface WatchSyncEffect {
    data object OpenConnection : WatchSyncEffect
    data object CloseConnection : WatchSyncEffect
    data object RefetchContinueWatching : WatchSyncEffect
    data object RefetchHistory : WatchSyncEffect
    data object RefetchWatchlist : WatchSyncEffect
    data object RefetchRatings : WatchSyncEffect
    data object RefetchHome : WatchSyncEffect
}

data class WatchSyncResult(
    val state: WatchSyncState,
    val effects: List<WatchSyncEffect>,
)

fun createWatchSyncState(profileId: String): WatchSyncState =
    WatchSyncState(
        profileId = profileId,
        connection = WatchSyncConnection.DISCONNECTED,
        isSurfaceVisible = false,
    )

/** Maps a server [WatchSyncKind] to the client refetch effects it triggers. */
private fun effectsForKind(kind: WatchSyncKind): List<WatchSyncEffect> =
    when (kind) {
        WatchSyncKind.ContinueWatching -> listOf(WatchSyncEffect.RefetchContinueWatching)
        WatchSyncKind.History ->
            listOf(WatchSyncEffect.RefetchContinueWatching, WatchSyncEffect.RefetchHistory)
        WatchSyncKind.Watchlist -> listOf(WatchSyncEffect.RefetchWatchlist)
        WatchSyncKind.Ratings -> listOf(WatchSyncEffect.RefetchRatings)
        WatchSyncKind.Home -> listOf(WatchSyncEffect.RefetchHome)
        WatchSyncKind.Unknown -> emptyList()
    }

fun reduceWatchSync(state: WatchSyncState, event: WatchSyncEvent): WatchSyncResult {
    return when (event) {
        WatchSyncEvent.SurfaceBecameVisible -> {
            val next = state.copy(isSurfaceVisible = true)
            if (state.connection == WatchSyncConnection.DISCONNECTED) {
                WatchSyncResult(
                    next.copy(connection = WatchSyncConnection.CONNECTING),
                    listOf(WatchSyncEffect.OpenConnection),
                )
            } else {
                WatchSyncResult(next, emptyList())
            }
        }

        WatchSyncEvent.SurfaceHidden -> {
            if (state.connection != WatchSyncConnection.DISCONNECTED) {
                WatchSyncResult(
                    state.copy(isSurfaceVisible = false, connection = WatchSyncConnection.DISCONNECTED),
                    listOf(WatchSyncEffect.CloseConnection),
                )
            } else {
                WatchSyncResult(state.copy(isSurfaceVisible = false), emptyList())
            }
        }

        WatchSyncEvent.ConnectionOpened -> {
            if (state.isSurfaceVisible) {
                WatchSyncResult(
                    state.copy(connection = WatchSyncConnection.CONNECTED),
                    listOf(
                        WatchSyncEffect.RefetchContinueWatching,
                        WatchSyncEffect.RefetchHome,
                    ),
                )
            } else {
                WatchSyncResult(
                    state.copy(connection = WatchSyncConnection.DISCONNECTED),
                    listOf(WatchSyncEffect.CloseConnection),
                )
            }
        }

        WatchSyncEvent.ConnectionClosed -> {
            WatchSyncResult(state.copy(connection = WatchSyncConnection.DISCONNECTED), emptyList())
        }

        is WatchSyncEvent.InvalidationReceived -> {
            if (state.connection == WatchSyncConnection.CONNECTED && event.profileId == state.profileId) {
                WatchSyncResult(state, effectsForKind(event.kind))
            } else {
                WatchSyncResult(state, emptyList())
            }
        }

        WatchSyncEvent.MaxDurationElapsed -> {
            if (state.connection != WatchSyncConnection.DISCONNECTED) {
                val effects = mutableListOf<WatchSyncEffect>(WatchSyncEffect.CloseConnection)
                if (state.isSurfaceVisible) {
                    effects += WatchSyncEffect.OpenConnection
                    WatchSyncResult(state.copy(connection = WatchSyncConnection.CONNECTING), effects)
                } else {
                    WatchSyncResult(state.copy(connection = WatchSyncConnection.DISCONNECTED), effects)
                }
            } else {
                WatchSyncResult(state, emptyList())
            }
        }
    }
}

fun WatchSyncEffect.toContractValue(): String =
    when (this) {
        WatchSyncEffect.OpenConnection -> "open_connection"
        WatchSyncEffect.CloseConnection -> "close_connection"
        WatchSyncEffect.RefetchContinueWatching -> "refetch_continue_watching"
        WatchSyncEffect.RefetchHistory -> "refetch_history"
        WatchSyncEffect.RefetchWatchlist -> "refetch_watchlist"
        WatchSyncEffect.RefetchRatings -> "refetch_ratings"
        WatchSyncEffect.RefetchHome -> "refetch_home"
    }

File: ios/ContractRunner/Sources/ContractRunner/WatchSync.swift
import Foundation

public enum WatchSyncConnection: Equatable {
    case disconnected
    case connecting
    case connected
}

public struct WatchSyncState: Equatable {
    public let profileId: String
    public let connection: WatchSyncConnection
    public let isSurfaceVisible: Bool

    public init(profileId: String, connection: WatchSyncConnection, isSurfaceVisible: Bool) {
        self.profileId = profileId
        self.connection = connection
        self.isSurfaceVisible = isSurfaceVisible
    }
}

/// The surface an invalidation targets. Sent by the server on the `watch_changed`
/// SSE event so the client can route each invalidation to the right refetch.
public enum WatchSyncKind: Equatable {
    case continueWatching
    case history
    case watchlist
    case ratings
    case home
    case unknown

    /// A legacy server omits `kind`; treat it as `continueWatching`.
    static func fromRaw(_ raw: String?) -> WatchSyncKind {
        switch raw?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "continue_watching": return .continueWatching
        case "history": return .history
        case "watchlist": return .watchlist
        case "ratings": return .ratings
        case "home": return .home
        case nil, "": return .continueWatching
        default: return .unknown
        }
    }
}

public enum WatchSyncEvent: Equatable {
    case surfaceBecameVisible
    case surfaceHidden
    case connectionOpened
    case connectionClosed
    case maxDurationElapsed
    case invalidationReceived(profileId: String, kind: WatchSyncKind, atMs: Int64)
}

public enum WatchSyncEffect: Equatable {
    case openConnection
    case closeConnection
    case refetchContinueWatching
    case refetchHistory
    case refetchWatchlist
    case refetchRatings
    case refetchHome

    public var contractValue: String {
        switch self {
        case .openConnection: return "open_connection"
        case .closeConnection: return "close_connection"
        case .refetchContinueWatching: return "refetch_continue_watching"
        case .refetchHistory: return "refetch_history"
        case .refetchWatchlist: return "refetch_watchlist"
        case .refetchRatings: return "refetch_ratings"
        case .refetchHome: return "refetch_home"
        }
    }
}

public struct WatchSyncResult: Equatable {
    public let state: WatchSyncState
    public let effects: [WatchSyncEffect]
}

public func createWatchSyncState(profileId: String) -> WatchSyncState {
    WatchSyncState(profileId: profileId, connection: .disconnected, isSurfaceVisible: false)
}

private func effects(for kind: WatchSyncKind) -> [WatchSyncEffect] {
    switch kind {
    case .continueWatching: return [.refetchContinueWatching]
    case .history: return [.refetchContinueWatching, .refetchHistory]
    case .watchlist: return [.refetchWatchlist]
    case .ratings: return [.refetchRatings]
    case .home: return [.refetchHome]
    case .unknown: return []
    }
}

public func reduceWatchSync(state: WatchSyncState, event: WatchSyncEvent) -> WatchSyncResult {
    switch event {
    case .surfaceBecameVisible:
        if state.connection == .disconnected {
            return WatchSyncResult(
                state: WatchSyncState(profileId: state.profileId, connection: .connecting, isSurfaceVisible: true),
                effects: [.openConnection]
            )
        }
        return WatchSyncResult(
            state: WatchSyncState(profileId: state.profileId, connection: state.connection, isSurfaceVisible: true),
            effects: []
        )

    case .surfaceHidden:
        if state.connection != .disconnected {
            return WatchSyncResult(
                state: WatchSyncState(profileId: state.profileId, connection: .disconnected, isSurfaceVisible: false),
                effects: [.closeConnection]
            )
        }
        return WatchSyncResult(
            state: WatchSyncState(profileId: state.profileId, connection: .disconnected, isSurfaceVisible: false),
            effects: []
        )

    case .connectionOpened:
        if state.isSurfaceVisible {
            return WatchSyncResult(
                state: WatchSyncState(profileId: state.profileId, connection: .connected, isSurfaceVisible: state.isSurfaceVisible),
                effects: [.refetchContinueWatching, .refetchHome]
            )
        }
        return WatchSyncResult(
            state: WatchSyncState(profileId: state.profileId, connection: .disconnected, isSurfaceVisible: state.isSurfaceVisible),
            effects: [.closeConnection]
        )

    case .connectionClosed:
        return WatchSyncResult(
            state: WatchSyncState(profileId: state.profileId, connection: .disconnected, isSurfaceVisible: state.isSurfaceVisible),
            effects: []
        )

    case let .invalidationReceived(profileId, kind, _):
        if state.connection == .connected && profileId == state.profileId {
            return WatchSyncResult(state: state, effects: effects(for: kind))
        }
        return WatchSyncResult(state: state, effects: [])

    case .maxDurationElapsed:
        if state.connection != .disconnected {
            if state.isSurfaceVisible {
                return WatchSyncResult(
                    state: WatchSyncState(profileId: state.profileId, connection: .connecting, isSurfaceVisible: state.isSurfaceVisible),
                    effects: [.closeConnection, .openConnection]
                )
            }
            return WatchSyncResult(
                state: WatchSyncState(profileId: state.profileId, connection: .disconnected, isSurfaceVisible: state.isSurfaceVisible),
                effects: [.closeConnection]
            )
        }
        return WatchSyncResult(state: state, effects: [])
    }
}
